doc/developers.rst

```diff
@@ -451,6 +444,138 @@ DataAccess class i.e. the `_field_write_arguments()` and
 `Arguments` class.
 
 
+Parsing Code
+############
+
+The PSyclone `parse` module is responsible for parsing science
+(algorithm and kernel) code and extracting the required information
+for the algorithm translation and PSy generation phases.
+
+The `parse` module contains modules for parsing algorithm
+(`algorithm.py`) and kernel (`kernel.py`) code as well as a utility
+module (`utils.py`) for common functionality.
+
+Parsing Algorithm Code
+======================
+
+The first thing PSyclone typically does is parse an input file. The
+code to do this is found in `parse/algorithm.py`.
+
+An input file can be parsed via the `parse` function or via an
+instance of the `Parser` class. In practice the `parse` function
+simply calls the `Parser` class so we will concentrate on the latter
+in this section. The `parse` function could be removed from PSyclone
+but it is simple and is used in existing PSyclone scripts and
+examples.
+
+The `Parser` class is initialised with a number of optional
+arguments. A particular `api` can be specified (this is required so
+the parser knows what sort of code and metadata to expect, how to
+parse it and which `builtins` are supported). The name used to specify
+an invoke (defaulting to `invoke`) can be changed, the name of the
+module supplying any infrastructure (defaulting to `inf`) can be
+changed, a path to where to look for associated kernel files can be
+provided and a particular maximum line length can be specified.
+
+Once an instance of the `Parser()` class is created and configured
+with required values for the optional arguments, then the parse method
+can be called. This takes the name of the input code as its argument
+and returns a parse tree of the input code and a FileInfo object that
+captures the required invoke information found in the input code and
+in the associated kernel codes.
+
+If the `nemo` API is specified then the `parse` method of the `Parser`
+instance simply parses the code with `fparser2` and returns the
+resultant `fparser2` parse tree.
+
+For all other APIs the `parse` method of the `Parser` instance returns
+the resultant `fparser2` parse tree and a `FileInfo` instance which
+captures the invoke and kernel metadata in a hierarchy of classes.
+
+When the `Parser` instance parses the code it expects to find Fortran
+code program, module, subroutine or function (and it aborts if
```

Comment:
"code containing either a program, module,..."?

Reply:
I've added containing but not either as we could have more than one in a file in theory.